lsm/main.cpp

```diff
@@ -5,31 +5,158 @@
+public:
+  void insert_key(int32_t k) {
+    auto key_hash = std::hash<int32_t>{}(k);
+    auto last_d = key_hash % 10;
+    bset[last_d] = 1;
+  }
+  char contains_key(int32_t k) {
+    auto key_hash = std::hash<int32_t>{}(k);
+    auto last_d = key_hash % 10;
+    return bset[last_d];
+  }
+};
+
+class File {
+public:
+  std::string filename;
+  int min, max;
+  BloomFilter bloomFilter;
+
+  File(std::string filename, BloomFilter bloomFilter, int min, int max) {
+    this->filename = filename;
+    this->bloomFilter = bloomFilter;
+    this->min = min;
+    this->max = max;
+  }
+};
+
+class maybe_value {
+public:
+  int32_t v;
+  char is_deleted;
+  explicit maybe_value() : v{0}, is_deleted{0} {}
+  explicit maybe_value(int32_t _v, char _deleted) : v{_v}, is_deleted{_deleted} {}
+};
 
 class LSM {
-  std::map<int32_t, int32_t> m;
+  std::map<int32_t, maybe_value> m;
+  std::vector<File> files;
 public:
-  void put(int32_t k, int32_t v) {
-    m.insert(std::make_pair(k,v));
+  void put(int32_t k, int32_t v, char deleted = 0) {
+    if (m.size() >= 4) {
+      std::string filename = std::to_string(files.size());
+      files.push_back(write_to_file(filename));
+      m.clear();
+    }
+    m.insert_or_assign(k,maybe_value(v, deleted));
   }
   std::optional<int32_t> get(int32_t k) {
     auto it = m.find(k);
+
     if (it != m.end()) {
-        return it->second;
+      if(!it->second.is_deleted) {
+        return it->second.v;
+      } else {
+        return std::nullopt;
+      }
+    } else {
+      for (auto i = files.rbegin(); i != files.rend(); ++i ) {
+        File file = *i;
+        if (file.bloomFilter.contains_key(k) && k >= file.min && k <= file.max) {
+          std::map<int32_t, maybe_value> file_map = read_from_file(file.filename);
+          auto it_m = file_map.find(k);
+          if (it_m != m.end()) {
+            if(!it_m->second.is_deleted) {
+              return it_m->second.v;
+            } else {
+              return std::nullopt;
+            }
+          }
+        }
+      }
     }
     return std::nullopt;
   }
   std::vector<std::pair<int32_t, int32_t>> range(int32_t l, int32_t r) {
     std::vector<std::pair<int32_t, int32_t>> res;
+    std::map<int32_t,int32_t>  res_map;
+
+    for (auto file : files) {
+      if (r >= file.min && l <= file.max) {
+        std::map<int32_t, maybe_value> file_map = read_from_file(file.filename);
+        auto it_ml = file_map.lower_bound(l);
+        auto it_mu = file_map.lower_bound(r);
+        for (auto it=it_ml; it!=it_mu; ++it) {
+          if (!it->second.is_deleted) {
+            res_map.insert_or_assign(it->first, it->second.v);
+          } else {
+            res_map.erase(it->first);
+          }
+        }
+      }
+    }
+
     auto it_l = m.lower_bound(l);
     auto it_u = m.lower_bound(r);
     for (auto it=it_l; it!=it_u; ++it) {
-        res.push_back(std::make_pair(it->first, it->second));
+      res_map.insert_or_assign(it->first, it->second.v);
```

Comment:
There can be deletes in the in-memory buffer as well:
```
p 1 3
p 10 1
p 11 1
p 12 1
d 1
r 0 9
```
should be an empty line

Reply:
In fact, I think this test will fail:
```
p 1 3
d 1
r 0 9
```
should be an empty line